scripts/main/frame.js

```diff
@@ -0,0 +1,221 @@
+};
+
+/**
+ * Initialized the DOM (if called for the very first time), sets the frontend
+ * into "frame mode" and enters the photo loading loop.
+ *
+ * @returns {void}
+ */
+frame.initAndStart = function () {
+	lychee.setMode("frame");
+	if (frame._dom.bgImage === null) {
+		frame._dom.bgImage = document.getElementById("background");
+		frame._dom.bgImage.addEventListener("load", function () {
+			// After a new background image has been loaded, draw a blurry
+			// version on the canvas.
+			StackBlur.image(frame._dom.bgImage, frame._dom.canvas, 20);
+			// We must reset the canvas to its originally defined dimensions
+			// as StackBlur reset its.
+			frame._dom.canvas.style.width = null;
+			frame._dom.canvas.style.height = null;
+		});
+	}
+	if (frame._dom.canvas === null) {
+		frame._dom.canvas = document.getElementById("background_canvas");
+	}
+	if (frame._dom.image === null) {
+		frame._dom.image = document.getElementById("picture");
+		frame._dom.image.addEventListener("load", function () {
+			// After a new image has been loaded, open the shutter
+			frame._dom.shutter.classList.add("opened");
+		});
+	}
+	if (frame._dom.shutter === null) {
+		frame._dom.shutter = document.getElementById("frame-shutter");
+	}
+
+	// We also must call the very first invocation of `runPhotoLoop`
+	// asynchronously to ensure that `nextTimeOutId` is also set for the first
+	// call, otherwise `frame.isRunning` and `frame.stop` will report false
+	// results and not work during the first invocation.
+	frame.nextTimeOutId = setTimeout(() => frame.runPhotoLoop(), 0);
+};
+
+/**
+ * Repeatedly loads random photos by calling {@link frame.loadRandomPhoto}
+ * every {@link lychee.mod_frame_refresh} interval forever unless an error
+ * occurs.
+ *
+ * The method stops loading photos upon an error or when {@link frame.stop}
+ * is called.
+ *
+ * @returns {void}
+ */
+frame.runPhotoLoop = function () {
+	/**
+	 * Recalls this method after the refresh timeout unless the loop
+	 * hasn't been stopped in the meantime.
+	 *
+	 * @returns {void}
+	 */
+	const onSuccess = function () {
+		if (frame.nextTimeOutId !== 0) {
+			frame.nextTimeOutId = setTimeout(() => frame.runPhotoLoop(), 1000 * lychee.mod_frame_refresh);
+		}
+	};
+
+	/**
+	 * Shows the error prominently and stops the loop.
+	 *
+	 * This is the old error handler when the frame mode was independently
+	 * implemented of the remaining frontend.
+	 * For historic reasons, the error is shown in a browser-provided
+	 * alert box.
+	 *
+	 * @param {XMLHttpRequest} jqXHR
+	 * @param {Object} params
+	 * @param {?LycheeException} lycheeException
+	 *
+	 * @returns {boolean}
+	 */
+	const onError = function (jqXHR, params, lycheeException) {
+		const msg = jqXHR.statusText + (lycheeException ? " - " + lycheeException.message : "");
+		console.error("The server returned an error response", {
+			description: msg,
+			params: params,
+			response: lycheeException,
+		});
+		alert(msg);
+		frame.stop();
```

Comment:
>  the alert box seems like a bad idea (I realize it was there before this refactoring).

I know exactly what you mean.

Yes, I kept it like this as it was this way before. I guess the frame mode used the alert box (instead of the disappearing error bar which Lychee provides), because the error bar was not available in frame mode until now. After https://github.com/LycheeOrg/Lychee-front/pull/335 has been merged, we can consider to use that error bar. Currently, it won't do anything good (even if it available now), because the background box of the frame mode still covers the error bar.

We could remove the alert box and add a "TODO" comment instead which hopefully remembers us to use the error bar once it will become possible or we leave it this way for now. I don't mind either way.

> I'm wondering if stopping the loop is the right thing to do?

For me this is the logical consequence of showing an alert box. The alert box will hide most of the image anyway and need to be clicked away explicitly. Hence, continuing the loop is not very helpful.

For me this is a both or nothing decision. If we keep the alert box, we should also stop the loop. If we remove the alert box, we can try going on.

@kamil4 It's your decision.

Reply:
I recommend to comment out the alert box, add a TODO, and let the loop continue.

Reply:
Fixed it in https://github.com/LycheeOrg/Lychee-front/pull/323/commits/3b3de969ef0764bbad4376a4d8ef38db2805a3dd.